### PR TITLE
Add more background for CS0825

### LINKED ISSUE #9102: Some background would be nice
Describing why this limitation exists would be a nice addition.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 77652ea7-f08b-6024-ec90-7899488cb988
* Version Independent ID: 70e8934d-9d82-acbc-f190-b7b6b9f230b3
* Content: [Compiler Error CS0825](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0825)
* Content Source: [docs/csharp/misc/cs0825.md](https://github.com/dotnet/docs/blob/master/docs/csharp/misc/cs0825.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by jmarkman**
## Summary

I went and digested Eric Lippert's [MSDN blog post](https://blogs.msdn.microsoft.com/ericlippert/2009/01/26/why-no-var-on-fields/) and [related StackOverflow post](https://stackoverflow.com/a/1929542) to address zspitz's request in #9102. With all my might, I had to resist calling the compiler's issue with `var` fields a "chicken and egg situation" and decided on calling it a logical paradox.

I also expanded upon the original code snippet slightly: I changed the variable names to be both friendlier and more concrete examples, and added a class field that had an expression on the right side of the field statement.

I wasn't sure if there were any policies in place for linking to external-yet-still-Microsoft sources, as I wanted to include a link to Eric's initial blog post either as an inline link, in **See Also**, or both. I refrained from including it until I received feedback. I also wasn't sure if I should have included a third field snippet that had an anonymous type, because I'd be lifting a lot from Eric's initial article and SO answer at that point.

Fixes #9102

### CODE REVIEW

File: docs/csharp/misc/cs0825.md [hunk 1]
```diff
@@ -10,7 +10,17 @@ ms.assetid: 49393d23-ec5f-4b44-a3fd-7e0a95ac0edd
...
+
+ ```csharp
+private var bookTitles;
+ ```
+`bookTitles` is a class field given the type `var`. As the field has no expression to evaluate, it is impossible for the compiler to infer what type `bookTitles` is supposed to be. In addition, adding an expression to the field (like you would for a local variable) is also insufficient:
+
+```csharp
+private var bookTitles = new List<string>();
+```
+When the compiler encounters fields during code compilation, it records each field's type before processing any expressions associated with it. The compiler encounters the same paradox trying to parse `bookTitles`: it needs to know the type of the field, but the compiler would normally determine `var`'s type by evaluating the expression, which isn't possible without knowing the type beforehand.
```
Comment (pkulikov) [inline docs/csharp/misc/cs0825.md]: Expressions are not evaluated by a compiler (except the constant ones). Above you use the "analyze" word in this context.

```suggestion
When the compiler encounters fields during code compilation, it records each field's type before processing any expressions associated with it. The compiler encounters the same paradox trying to parse `bookTitles`: it needs to know the type of the field, but the compiler would normally determine `var`'s type by analyzing the expression, which isn't possible without knowing the type beforehand. 
```

However, then the last part of the sentence becomes wrong.

@jmarkman I think you caught everything in this sentence:
>When the compiler encounters fields during code compilation, it records each field's type before processing any expressions associated with it.

Then logical paradox is clear:
 1. To infer type, the compiler needs to analyze the initialization expression.
 2. The compiler requires the field types to be known before checking any initialization expression.

**Comment by BillWagner** [comment]
Thanks for submitting this @jmarkman 

I need to think about this more before proposing what to do next. What you wrote is good, but I have some concerns:
1. **Is this the best location for the extended explanation?** Our goal on the compiler warning and error pages is to give a reader the information they need to fix the error. These are tuned to search well on the error code and message, but not on the concepts.  This content might be better in a different location, and linked from this topic.
2. **How is this affected by the proposed [target typed new](https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-new.md) proposal?** This would really change both the explanation and the fix for this issue. Note that this proposal is not scheduled for any release at this time.

I will post again on Monday.

**Comment by jmarkman** [comment]
For consideration on Monday: in regards to 1, should that just be a default response to queries of #9102 's nature (i.e., asking for extended background on compiler warning & error pages)? I picked out the issue because I was curious myself after seeing a question about it pop up elsewhere. 

@pkulikov does the change to "analyze" make the last part of the sentence wrong? I'm not well versed in regards to compilers in general outside of their utmost basic functions, so I relied heavily on Eric Lippert's explanation during my interpretation.

**Comment by pkulikov** [comment]
@jmarkman as far as I've noticed, when docs say "evaluate an expression", they mean to perform all necessary methods and operator calls to compute and return the result of the expression. I guess, that doesn't happen when the compiler infers the type of `var`. That's why my suggestion was to change the last sentence into (the only difference is to use "analyze" instead of "evaluate"):
>...it needs to know the type of the field, but the compiler would normally determine `var`'s type by analyzing the expression, which isn't possible without knowing the type beforehand. 

However, it's possible to analyze an expression without knowing its type, so the "...by analyzing the expression, which isn't possible without knowing the type beforehand." part becomes not correct. 

Thanks for the link to the blog post! (By the way, notice that it doesn't use the word "evaluate" :) )

**Review by BillWagner — DISMISSED** [review]
Thanks for submitting this explanation @jmarkman 

I did some research, and I have a better suggestion for where this new information should go. You've added important info, and it should be seen by a wider audience.

The pages for warnings and errors are tuned to provide good SEO only for the warning or error. The programming guide is tuned for the concepts that developers need. As a result, the page I'm suggesting you modify has 100X more views in the past month.

This page should link to the new content on the suggested page.

File: docs/csharp/misc/cs0825.md [hunk 2]
```diff
@@ -10,7 +10,17 @@ ms.assetid: 49393d23-ec5f-4b44-a3fd-7e0a95ac0edd
 # Compiler Error CS0825
 The contextual keyword 'var' may only appear within a local variable declaration.  
   
- Implicit typing with the `var` keyword can only be applied to variables at local method scope.  
+ Implicit typing with the `var` keyword can only be applied to variables at local method scope. Implicit typing is not available for class fields as the C# compiler would encounter a logical paradox as it processed the code: the compiler needs to know the type of the field, but it cannot determine the type until the assignment expression is analyzed, and the expression cannot be evaluated without knowing the type. Consider the following code:
```
[the remark this answers is not in the document]
Comment (jmarkman) [reply docs/csharp/misc/cs0825.md]: Moved!

**Comment by mairaw** [comment]
@jmarkman sorry that we lost track of this PR during the holidays. Given that a [PR](https://github.com/dotnet/docs/pull/9925) was merged since then that reformatted this whole topic, we now have some major merge conflicts. It might be easier to grab the new version and past your changes in. Let us know if you need any help with that.

/cc @BillWagner

**Comment by jmarkman** [comment]
Wow, this was a relative while ago. I'll take a look and reacquaint myself since I ended up taking a break.

**Comment by jmarkman** [comment]
I just pulled the new version in and re-applied my changes, hopefully the merge isn't unbearably messy.

**Review by BillWagner — APPROVED** [review]
Thanks for the extra updates @jmarkman This is ready to :shipit:

I've approved this, and I'll merge it into a staging branch. The changes will go live next week, during Build.

**Pull request merged**